pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/TopicStatsImpl.java

```diff
@@ -252,19 +253,15 @@ public TopicStatsImpl add(TopicStats ts) {
                    return newStats;
                }).add((PublisherStatsImpl) s);
            } else {
-               if (this.publishers.size() != stats.publishers.size()) {
-                   for (int i = 0; i < stats.publishers.size(); i++) {
-                       PublisherStatsImpl newStats = new PublisherStatsImpl();
-                       newStats.setSupportsPartialProducer(false);
-                       this.publishers.add(newStats.add(stats.publishers.get(i)));
-                   }
-               } else {
-                   for (int i = 0; i < stats.publishers.size(); i++) {
-                       this.publishers.get(i).add(stats.publishers.get(i));
-                   }
+               if (index == this.publishers.size()) {
+                   PublisherStatsImpl newStats = new PublisherStatsImpl();
+                   newStats.setSupportsPartialProducer(false);
+                   this.publishers.add(newStats);
                }
+               this.publishers.get(index)
+                       .add((PublisherStatsImpl) s);
```

Comment:
Hi,

I see that `*TopicStatsImpl` is not a thread-safe class. Currently, the caller needs to externally synchronize the access if concurrent threads access the same instance.

> Potentially access to it from concurrent threads.


Could you provide this example? I see that `PartitionedTopicStatsImpl` is instantiated by a single thread, and `add(..) ` gets called from the same thread.


```
PartitionedTopicStatsImpl stats = new PartitionedTopicStatsImpl(partitionMetadata);
...
            FutureUtil.waitForAll(topicStatsFutureList).handle((result, exception) -> {
...
                            stats.add(statFuture.get());
                            if (perPartition) {
                                stats.getPartitions().put(topicName.getPartition(i).toString(), statFuture.get());
...
                    }
                }
```

Reply:
This case may not be in the current codes. If contributors already know "the caller needs to externally synchronize the access," we don't care in this section. I'm worried about misimplementation. So, I think it's okay to write comments about the current spec instead of fixing it to follow my comment.

Reply:
Hi, sorry. Do you mean we need to comment that `PartitionedTopicStatsImpl` is not thread-safe?

Reply:
> Do you mean we need to comment that PartitionedTopicStatsImpl is not thread-safe?

Yes.

Reply:
Added comments.